src/app/MessageDef.h

```diff
@@ -0,0 +1,1196 @@
+     */
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter);
+};
+
+/**
+ *  @brief
+ *    CHIP interaction model Attribute Path definition
+ *
+ */
+namespace AttributePath {
+enum
+{
+    kCsTag_NodeId              = 0,
+    kCsTag_EndpointId          = 1,
+    kCsTag_NamespacedClusterId = 2,
+    kCsTag_FieldId             = 3,
+    kCsTag_ListIndex           = 4,
+};
+
+/**
+ *  @class AttributePath::Parser
+ *
+ *  @brief
+ *    Parses a Attribute Path container
+ */
+class Parser : public ParserBase
+{
+public:
+    // aReader has to be on the element of AttributePath container
+    CHIP_ERROR Init(const chip::TLV::TLVReader & aReader);
+
+    // Roughly verify the schema is right, including
+    // 1) all mandatory tags are present
+    // 2) no unknown tags
+    // 3) all elements have expected data type
+    // 4) any tag can only appear once
+    CHIP_ERROR CheckSchemaValidity() const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetNodeId(chip::NodeId * const apNodeId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetEndpointId(chip::EndpointId * const apEndpointId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetNamespacedClusterId(chip::ClusterId * const apClusterId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetFieldId(uint8_t * const apFieldId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetListIndex(uint16_t * const apListIndex) const;
+};
+
+/**
+ *  @class AttributePath::Builder
+ *
+ *  @brief
+ *    Attribute Path Builder
+ */
+class Builder : public BuilderBase
+{
+public:
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter);
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter, const uint8_t aContextTagToUse);
+
+    AttributePath::Builder & NodeId(const chip::NodeId aNodeId);
+    AttributePath::Builder & EndpointId(const chip::EndpointId aEndpointId);
+    AttributePath::Builder & NamespacedClusterId(const chip::ClusterId aNamespacedClusterId);
+    AttributePath::Builder & FieldId(const uint8_t aFieldId);
+    AttributePath::Builder & ListIndex(const uint16_t aListIndex);
+
+    AttributePath::Builder & EndOfAttributePath();
+
+private:
+    CHIP_ERROR _Init(chip::TLV::TLVWriter * const apWriter, const uint64_t aTag);
+};
+
+}; // namespace AttributePath
```

Comment:
nit: no need for ; after namespaces. Here and below as well.

Reply:
Done, thanks